app/src/main/java/ch/epfl/sweng/SDP/home/LeaderboardActivity.java

```diff
@@ -0,0 +1,378 @@
+         * Processes inquiry locally if cache (allPlayers) is not empty.
+         * Else fetches data from Firebase and stores it in allPlayers.
+         * @param query new string to search
+         */
+        private void update(String query) {
+            if (!allPlayers.isEmpty()) {
+                filterWantedPlayers(query);
+                Collections.sort(wantedPlayers);
+                leaderboardView.removeAllViews();
+                addWantedPlayersToLayout();
+            } else {
+                fetchPlayersFromFirebase(query);
+            }
+        }
+
+        /**
+         * Copies all players that contain query into wantedPlayers.
+         * @param query string to search
+         */
+        private void filterWantedPlayers(String query) {
+            wantedPlayers.clear();
+            for (Player tempPlayer : allPlayers) {
+                if (tempPlayer.playerNameContainsString(query)) {
+                    wantedPlayers.add(tempPlayer);
+                }
+            }
+        }
+
+        /**
+         * Gets called when local cache of players is empty.
+         * Adds snapshots to a list, filters it by query,
+         * sorts it by trophies and adds players to LinearLayout
+         *
+         * @param query string to search
+         */
+        private void fetchPlayersFromFirebase(final String query) {
+            Database.INSTANCE.getReference("users")
+                    .addListenerForSingleValueEvent(new ValueEventListener() {
+                        @Override
+                        public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
+                            for (DataSnapshot s : dataSnapshot.getChildren()) {
+                                if (s.child("userId") == null || s.child("username") == null
+                                        || s.child("trophies") == null) {
+                                    continue;
+                                }
+                                Player temp = new Player((String) s.child("userId").getValue(),
+                                        (String) s.child("username").getValue(),
+                                        (Long) s.child("trophies").getValue());
+
+                                allPlayers.add(temp);
+                            }
+                            filterWantedPlayers(query);
+                            Collections.sort(wantedPlayers);
+                            leaderboardView.removeAllViews();
+                            addWantedPlayersToLayout();
+                        }
+
+                        @Override
+                        public void onCancelled(@NonNull DatabaseError databaseError) {
+                            Log.d(TAG, FIREBASE_ERROR);
+                        }
+                    });
+        }
+
+        /**
+         * Adds wantedPlayers to the leaderboardView.
+         */
+        public void addWantedPlayersToLayout() {
```

Comment:
If I understand correct this function updates the UI. Therefore, I think it would be better if you put this method in the LeaderboardActivity class and add an argument **wantedPlayers** so it's separated from the database logic.

Reply:
There is always a trade-off between convention and practicability. In this case it make everything much more complicated because i need to call it once firebase has asynchronously sent all data, not before and not after.

Reply:
Okey, I agree with you. I believe this issue can be resolved then.